lib/src/rules/prefer_int_literals.dart

```diff
@@ -67,20 +67,30 @@ class _Visitor extends SimpleAstVisitor<void> {
 
     // Ensure that replacing the double would not change the semantics
     if (isDoubleContext(node)) {
-      rule.reportLintForToken(node.literal);
+      rule.reportLint(node);
     }
   }
 
   bool isDoubleContext(AstNode child) {
     final node = child.parent;
     if (node == null) {
       return false;
+    } else if (node is ArgumentList) {
+      if (child is Expression) {
+        return child.staticParameterElement?.type?.name == 'double';
+      }
+      return false;
+    } else if (node is AssignmentExpression) {
+      return node.rightHandSide?.staticType?.name == 'double';
+    } else if (node is BinaryExpression) {
```

Comment:
I'm having a hard time convincing myself that the logic for binary expressions is correct in the face of user defined operators. Consider
```
class C {
  double operator +(double value) => value;
}
void f(C c) {
  print(c + 2.0);
}
```
I think that should lint because the right operand would need to be auto converted based on the signature of the operator. But `operand` would be `c`, which is not a double, so I think this method would return `false`.

I think for user defined operators we really need to look at the type of the operator's parameter.

I'd have to look at each of the non-user-definable operators to decide whether this is right in those cases. We should probably have tests for the rest of the operators (though one test per operator is probably sufficient).

Reply:
Good point. I have revised the logic and added more tests. I've removed support for linting double literals in expressions until I can better understand the various situations.